log/loki.go

```diff
@@ -128,15 +163,47 @@ func LokiFromConfigLine(ctx context.Context, fallbackLogger logrus.FieldLogger,
 				continue
 			}
 
-			return nil, fmt.Errorf("unknown loki config key %s", key)
+			return fmt.Errorf("unknown loki config key %s", key)
 		}
 	}
 
-	h.client = &http.Client{Timeout: h.pushPeriod}
+	return nil
+}
 
-	go h.loop()
+func parseArray(key, value string, i int, args []string) ([]string, int, error) {
+	if value == "" {
+		return []string{}, i, nil
+	}
+	var ended bool
+	var result []string
+	if value[0] != '[' {
+		return nil, 0, fmt.Errorf("%s needs to be in the form [value1,value2,...,valueN]", key)
+	}
+	if value[len(value)-1] == ']' {
+		value = value[:len(value)-1]
+		ended = true
+	}
+	value = value[1:]
+
+	result = append(result, value)
+	for i++; i < len(args) && !ended; i++ {
+		value = args[i]
+		if value == "" {
+			return nil, 0, fmt.Errorf("array item in %s need to not be empty", key)
+		} else if value[len(value)-1] == ']' {
+			value = value[:len(value)-1]
+			ended = true
```

Comment:
This parsing seems convoluted to me... Why do you need to go over `args` at all? Wouldn't stripping the square brackets and doing `strings.Split(value, ",")` accomplish the same thing?

Reply:
I can't because I have already `args = strings.Split(line, ",")`  early on .. so there are no more `,` in the values ;).
The alternative to this was to rewrite the whole code to read byte by byte and when it gets to `,` or `[` to change what I do or something like that. This will probably be a great idea when we rewrite it to be used everywhere, but for now, it seems like too much work.

Reply:
Aha, right. Though a scanner implementation will probably be simpler than this, and shouldn't be too much work. I would prefer if we do it now instead of merging this, but I leave it up to you.

Reply:
fixed in 15fcb34

Reply:
Dude, how is that simpler? 😆 It's less hacky, I suppose, but it looks more complex than before. Why not use [`bufio.Scanner`](https://golang.org/pkg/bufio/#Scanner) for this? I was picturing this would be much simpler with that approach.

Reply:
I tried using [strings.Reader](https://golang.org/pkg/strings/#Reader) ( bufio.Scanner doesn't really help in any way :D). But in the end, I needed to add runes to each other to build a string to get the final keys and values out of one rune at a time (there is no "which index are we at") ... which seemed terrible and still had more issues, so I dropped it for the simplest scanning of one symbol at a time.

I don't think it's simpler but ti let's build on top of it to have `"values with , inside"` and `{dict:values}` and such.

Reply:
OK, if you think this is an improvement I'm fine with it, but I'm not sure having a custom parser for this is a good idea.

Reply:
well .. you didn't like the previous less custom ... we both don't like the strvals array syntax of `array[0]=2,array[2]=1` meaning `array=["2","","1"]`... so I don't see another option :)